core/src/main/java/org/elasticsearch/common/util/concurrent/KeyedLock.java

```diff
@@ -50,48 +52,37 @@ public KeyedLock() {
 
     private final ConcurrentMap<T, KeyLock> map = ConcurrentCollections.newConcurrentMap();
 
-    protected final ThreadLocal<KeyLock> threadLocal = new ThreadLocal<>();
-
-    public void acquire(T key) {
+    public Releasable acquire(T key) {
         while (true) {
-            if (threadLocal.get() != null) {
```

Comment:
I think moving to the releasable model makes it way less likely to happen by accident. I also feel it makes this class simpler to reason about and more in line with the fact it uses re-entrant locks. Any thread local removal is good, right? 

If we really want to protect against double locking from the same thread by mistake - maybe just calling isHeldByCurrentThread enough? potentially just as an assertion?